common/schema-builder/src/main/java/io/smallrye/graphql/schema/creator/OperationCreator.java

```diff
@@ -130,36 +132,116 @@ private void checkWrappedTypeKotlinNullability(MethodInfo methodInfo,
+        boolean nullable = Flag.Type.IS_NULLABLE.invoke(flags);
+        return nullable;
+    }
+
     private boolean compareParameterLists(List<KmValueParameter> kotlinParameters,
             List<Type> jandexParameters) {
         if (kotlinParameters.size() != jandexParameters.size()) {
             return false;
         }
         for (int i = 0; i < kotlinParameters.size(); i++) {
-            // TODO: the matching of parameter types could use some improvements
-            // For example, it won't work for primitives.
-            // An Int parameter will be represented as kotlin.Int in the KotlinClassMetadata,
-            // but as "int" in the Jandex MethodInfo.
-            if (!((KmClassifier.Class) kotlinParameters.get(i).getType().classifier)
-                    .getName().replace("/", ".")
-                    .equals(jandexParameters.get(i).name().toString())) {
+            KmType kotlinType = kotlinParameters.get(i).getType();
+            Type jandexType = jandexParameters.get(i);
+            if (!compareJavaAndKotlinType(jandexType, kotlinType)) {
                 return false;
             }
         }
         return true;
     }
 
+    private boolean compareJavaAndKotlinType(Type javaType, kotlinx.metadata.KmType kotlinType) {
+        if (kotlinType == null) {
+            return false;
+        }
+        String kotlinTypeName = ((KmClassifier.Class) kotlinType.classifier).getName().replace("/", ".");
+
+        boolean signatureIsEqual = false;
+        if (kotlinTypeName.equals(javaType.name().toString())) {
+            signatureIsEqual = true;
+        } else if (Classes.isCollection(javaType)) {
+            signatureIsEqual = compareKotlinCollectionType(javaType, kotlinType);
+        }
+        // TODO: the matching of parameter types could use some improvements
+        // TODO: only some cases are handled, there are much more: see https://kotlinlang.org/docs/java-interop.html#mapped-types
+
+        boolean parametersAreEqual = true;
+        if (javaType instanceof ParameterizedType) {
+            List<Type> arguments = javaType.asParameterizedType().arguments();
+            for (int i = 0; i < arguments.size(); i++) {
+                Type javaArg = arguments.get(0);
+                KmType kotlinArg = kotlinType.getArguments().get(i).getType();
+
+                boolean argTypeEqual = compareJavaAndKotlinType(javaArg, kotlinArg);
+                if (!argTypeEqual) {
+                    parametersAreEqual = false;
+                    break;
+                }
+            }
+        }
+
+        return signatureIsEqual && parametersAreEqual;
+    }
+
+    private boolean compareKotlinCollectionType(Type javaType, KmType kotlinType) {
+        String javaCollectionType = javaType.name().toString();
+        String kotlinCollectionType = ((KmClassifier.Class) kotlinType.classifier).getName().replace("/", ".");
+        // TODO: naive implementation. Mutable Types are relevant?
```

Comment:
Yeah I think the equals checks should also accept the mutable collection types, otherwise in `MutableList<Foo>` the `Foo` is detected as nullable